clinica/utils/filemanip.py

```diff
@@ -469,3 +469,74 @@ def extract_metadata_from_json(json_file: str, list_keys: List[str]) -> List[str
         raise ClinicaException(error_message)
 
     return list_values
+
+
+def select_values_from_metadata(dict_values: dict, json_file: str) -> list:
+    """This function selects the data needed for dwi-preprocessing-using-t1: TotalReadoutTime and PhaseEncodinDirection.
+    If it is not there, it looks for alternative metrics which work as well. If they're not found, the functions errors.
+
+    Parameters
+    ----------
+    dict_values: dict
+    Requested values dictionnary.
+
+    json_file: str
+    Path to a json file containing metadata needed for the pipeline.
+
+    Returns
+    -------
+    list: contains the values for the TotalReadoutTime and the PhaseEncodingDirection
+    """
+    from clinica.utils.exceptions import ClinicaException
+
+    if dict_values["TotalReadoutTime"] != "":
+        readouttime = dict_values["TotalReadoutTime"]
+    elif dict_values["EstimatedTotalReadoutTime"] != "":
+        readouttime = dict_values["EstimatedTotalReadoutTime"]
+    elif (
+        dict_values["PhaseEncodingSteps"] != "" and dict_values["PixelBandwidth"] != ""
+    ):
+        readouttime = dict_values["PhaseEncodingSteps"] / dict_values["PixelBandwidth"]
+    else:
+        raise ClinicaException(
+            f"Error: Clinica could not find the TotalReadoutTime or data to compute it in the following json file: {json_file}"
+        )
+    if dict_values["PhaseEncodingDirection"] != "":
+        phaseencodingdirection = dict_values["PhaseEncodingDirection"]
+    elif dict_values["PhaseEncodingAxis"] != "":
+        phaseencodingdirection = dict_values["PhaseEncodingAxis"] + "+"
+    else:
+        raise ClinicaException(
+            f"Error: Clinica could not find the PhaseEcondingDirection or the PhaseEncodingAxis in the following json file: {json_file}"
+        )
+    return [readouttime, phaseencodingdirection]
+
+
+def extract_metadata_from_dwi_json(json_file: str, list_keys: list) -> list:
```

Comment:
After some thinking, I don't believe we should have both `extract_metadata_from_json` and `extract_metadata_from_dwi_json`.
The former is only used for DWI and doesn't do anything specific to DWI, it just reads a JSON file, tries to filter the resulting dictionary with a user-provided list of keys, and raises if there are some missing keys. The only thing you want to change here is the way the missing keys are handled: you do not raise immediately, but you try to find alternative data, and raise if this is too desperate.

Also, the current API is a bit strange, since you provide a list of keys to the extract function and you get a list of different size in return.

So, I would remove this function entirely, and rely on `extract_metadata_from_json`. Below is a suggestion of a possible implementation (might contain errors as I wrote it directly here without trying it....) which should solve some of these issues. Tell me what you think, I might have misunderstood what you want to do here.

```python
def extract_metadata_from_json(
        json_file: str,
        list_keys: List[str],
        handle_missing_keys: Optional[Callable] = None,
) -> List[str]:
    """Same function but with additional parameter.
    Would need to update callers but they look limited.
    """
    data = read_json(json_file)  # replace with JSON reading code
    missing_keys = set(list_keys).difference(set(data.keys()))
    if len(missing_keys) > 0 and handle_missing_keys is None:
        raise ClinicaException(f"bla bla {missing_keys}.")
    patch = handle_missing_keys(data, missing_keys)
    data.update(patch)
    return [data[k] for k in list_keys]


def _handle_missing_keys_dwi(data: dict, missing_keys: set) -> dict:
    handlers = {
        "TotalReadoutTime": _handle_missing_total_readout_time,
        "PhaseEncodingDirection": _handle_missing_phase_encoding_direction,
    }
    try:
        return {k: handlers[k](data, missing_keys) for k in missing_keys}
    except KeyError:
        raise ValueError(f"Could not recover the missing keys {missing_keys} from JSON file.")


def _handle_missing_total_readout_time(data: dict, missing_keys: set) -> float:
    if "EstimatedTotalReadoutTime" in missing_keys:
        if "PhaseEncodingSteps" in missing_keys and "PixelBandwidth" in missing_keys:
            raise ClinicaException("bla bla bla")
        else:
            return data["PhaseEncodingSteps"] / data["PixelBandwidth"]  # What about zero div errors here ?
    else:
        return data["EstimatedTotalReadoutTime"]


def _handle_missing_phase_encoding_direction(data: dict, missing_keys: set) -> float:
     # Same philosophy here

```

Then you would call it in the following way:

```python
[total_readout_time, phase_encoding_direction] = extract_metadata_from_json(
         dwi_json,
        ["TotalReadoutTime", "PhaseEncodingDirection"],
        handle_missing_keys=_handle_missing_keys_dwi,
)
```

Reply:
Alright ! Makes sense to me, i'll try to implement it along the lines you are suggesting !